### PR TITLE
ITE drivers/interrupt_controller: add wuc interface

### PR DESCRIPTION
**Pull request by RuibinChang**
Add wakeup controller interface for ITE it8xxx2 chip.

Signed-off-by: Ruibin Chang [ruibin.chang@ite.com.tw](mailto:ruibin.chang@ite.com.tw)

### CODE REVIEW

**Comment by RuibinChang** [comment]
Sorry, because PR#41563 is nothing changed on github website when everytime I push the commit, I create this PR which is mirrored from PR#41563 and include following changes:
1.rebase
2.solve kconfig conflict
3.change the deprecated UTIL_LISTIFY macro to LISTIFY

After this PR is merged, I will close PR#41563.

**Review by mbolivar-nordic — DISMISSED** [review]

File: drivers/interrupt_controller/wuc_ite_it8xxx2.c [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+};
+
+void it8xxx2_wuc_enable(const struct device *dev, uint8_t mask)
+{
+	const struct it8xxx2_wuc_cfg *config = dev->config;
+	volatile uint8_t *reg_wuenr = config->reg_wuenr;
+
+	/*
+	 * WUC group only 1, 3, and 4 have enable/disable register,
+	 * others are always enable.
```
Comment (fkokosinski) [inline drivers/interrupt_controller/wuc_ite_it8xxx2.c:39]: Typo? `enable` -> `enabled`

File: drivers/interrupt_controller/wuc_ite_it8xxx2.c [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+void it8xxx2_wuc_disable(const struct device *dev, uint8_t mask)
+{
+	const struct it8xxx2_wuc_cfg *config = dev->config;
+	volatile uint8_t *reg_wuenr = config->reg_wuenr;
+
+	/*
+	 * WUC group only 1, 3, and 4 have enable/disable register,
+	 * others are always enable.
```
Comment (fkokosinski) [inline drivers/interrupt_controller/wuc_ite_it8xxx2.c:56]: Ditto

**Review by fkokosinski — COMMENTED** [review]
Only minor typos, the rest looks good.

File: drivers/kscan/kscan_ite_it8xxx2.c [hunk 3]
```diff
@@ -19,6 +21,8 @@
 #define LOG_LEVEL CONFIG_KSCAN_LOG_LEVEL
 LOG_MODULE_REGISTER(kscan_ite_it8xxx2);
 
+#define KEYBOARD_KSI_PIN_COUNT		0x8
```
Comment (keith-zephyr) [inline drivers/kscan/kscan_ite_it8xxx2.c:24]: This count can be determined by the `IT8XXX2_DT_INST_WUCCTRL_LEN(0)`.

File: drivers/kscan/kscan_ite_it8xxx2.c [hunk 4]
```diff
@@ -19,6 +21,8 @@
...
+#define KEYBOARD_KSI_PIN_MASK		0xff
```
Comment (keith-zephyr) [inline drivers/kscan/kscan_ite_it8xxx2.c:25]: This macro assumes that all KSI pins are connected to the same wake up controller.  I'm not sure if it's possible to create build assert to verify this.  You could add a run-time to check to `kscan_it8xxx2_init()`.

You should also set the KS pin mask value from the devicetree.  You can do this at runtime as shown below, or you can probably use LISTIFY() to create the mask at compile time.

```
const struct device *kscan_wucs = config->wuc_map_list[0].wucs;
config->ksi_pin_mask = 0;
for (int i = 0; i < KEYBOARD_KSI_PIN_COUNT; i++) {
    if (IS_ENABLED(CONFIG_LOG)) {
        if (config->wuc_map_list[i].wucs != kscan_wucs) {
            LOG_ERR(...);
        }
    }
    config->ksi_pin_mask |= config->wuc_map_list[i].mask;

```

**Review by keith-zephyr — CHANGES_REQUESTED** [review]

File: drivers/interrupt_controller/wuc_ite_it8xxx2.c (hunk 1 above)
Comment (RuibinChang) [reply drivers/interrupt_controller/wuc_ite_it8xxx2.c]: Done

File: drivers/interrupt_controller/wuc_ite_it8xxx2.c (hunk 2 above)
Comment (RuibinChang) [reply drivers/interrupt_controller/wuc_ite_it8xxx2.c]: Done

File: soc/riscv/riscv-ite/common/soc_dt.h [hunk 5]
```diff
@@ -41,4 +41,61 @@
...
+ *
+ * @param inst instance number
+ * @return length of wucctrl property which type is 'phandle-array'
+ */
+#define IT8XXX2_DT_INST_WUCCTRL_LEN(inst)				\
+	DT_INST_PROP_LEN(inst, wucctrl)
+
+/**
+ * @brief For it8xxx2, construct an array of it8xxx2 wuc map structure
+ *        with compatible defined in DT_DRV_COMPAT by UTIL_LISTIFY func
```
[the remark this answers is not in the document]
Comment (RuibinChang) [reply soc/riscv/riscv-ite/common/soc_dt.h]: Done

File: drivers/kscan/kscan_ite_it8xxx2.c (hunk 3 above)
Comment (RuibinChang) [reply drivers/kscan/kscan_ite_it8xxx2.c]: Done

File: drivers/kscan/kscan_ite_it8xxx2.c (hunk 4 above)
Comment (RuibinChang) [reply drivers/kscan/kscan_ite_it8xxx2.c]: Done, but I move ksi_pin_mask to data struct, since it's not a property.

**Review by keith-zephyr — APPROVED** [review]

**Comment by RuibinChang** [comment]
Are there any other comments?

**Comment by Dino-Li** [comment]
@jackrosenthal @brockus-zephyr @mbolivar-nordic can you re-review the change?

**Review by brockus-zephyr — APPROVED** [review]

**Pull request merged**